File: src/main/java/io/github/hhy/bookmark/manager/HighVersionBookManager.java
package io.github.hhy.bookmark.manager;

import com.intellij.ide.bookmark.*;
import com.intellij.openapi.project.Project;
import io.github.hhy.bookmark.element.Element;
import io.github.hhy.bookmark.element.ElementBuilder;
import io.github.hhy.bookmark.element.Type;
import io.github.hhy.bookmark.notify.Notify;
import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.lang.StringUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class HighVersionBookManager implements MyBookmarkManager {

    @Override
    public List<Element> getAllBookmarks(Project project) {
        List<Element> result = new ArrayList<>();
        if (project.getService(BookmarksManager.class) instanceof BookmarksManagerImpl bookmarkManager) {
            ManagerState managerState = bookmarkManager.getState();
            for (GroupState group : managerState.getGroups()) {
                Element gEle = Element.createGroup(group.getName());
                result.add(gEle);

                for (BookmarkState bookmark : group.getBookmarks()) {
                    Map<String, String> attr = bookmark.getAttributes();
                    Element bEle = ElementBuilder.anElement()
                            .withElementType(Type.BOOKMARK)
                            .withName(bookmark.getDescription())
                            .withFileDescriptor(Optional.ofNullable(attr.get("url")).map(HighVersionBookManager::urlToFileDescriptor).orElseGet(() -> ""))
                            .withLinenumber(Optional.ofNullable(attr.get("line")).map(Integer::parseInt).orElseGet(() -> -1))
                            .withGroup(group.getName())
                            .withBookmarkType(bookmark.getType().toString())
                            .build();
                    result.add(bEle);
                }
            }
        }
        return result;
    }

    @Override
    public void addBookmarks(Project project, List<Element> elements) {
        if (CollectionUtils.isEmpty(elements)) return;
        BookmarksManager bookmarksManager = project.getService(BookmarksManager.class);
        for (Element element : elements) {
            try {
                switch (element.elementType()) {
                    case GROUP -> addGroup(bookmarksManager, element.name());
                    case BOOKMARK -> {
                        BookmarkGroup group = addGroup(bookmarksManager, element.group());
                        addBookmark(bookmarksManager, group, element);
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
                Notify.error(e.getMessage());
            }
        }
    }

    @Override
    public Element getBookmark(Project project, String fileDescription, int lineNumber) {
        for (Element element : getAllBookmarks(project)) {
            if (element.elementType() != Type.BOOKMARK) continue;
            if (element.fileDescriptor().equals(fileDescription) && element.linenumber() == lineNumber) {
                return element;
            }
        }
        return null;
    }

    @Override
    public List<Element> removeInvalid(Project project) {
        List<Element> invalid = new ArrayList<>();
        BookmarksManager bookmarksManager = project.getService(BookmarksManager.class);
        for (Bookmark bookmark : bookmarksManager.getBookmarks()) {
            if (bookmark.getClass().getName().contains("InvalidBookmark")) {
                bookmarksManager.remove(bookmark);

                Map<String, String> attr = bookmark.getAttributes();
                Element ele = Element.createBookmark(Optional.ofNullable(attr.get("url")).map(HighVersionBookManager::urlToFileDescriptor).orElseGet(() -> ""),
                        Optional.ofNullable(attr.get("line")).map(Integer::parseInt).orElseGet(() -> -1));
                invalid.add(ele);
            }
        }
        return invalid;
    }

    private BookmarkGroup addGroup(BookmarksManager bookmarksManager, String groupName) {
        BookmarkGroup group = bookmarksManager.getGroup(groupName);
        if (group == null) {
            return bookmarksManager.addGroup(groupName, false);
        }
        return group;
    }

    private void addBookmark(BookmarksManager bookmarksManager, BookmarkGroup group, Element element) {
        if (StringUtils.isNotEmpty(element.fileDescriptor())) {
            BookmarkType type = BookmarkType.valueOf(element.bookmarkType());

            BookmarkState bookmarkState = new BookmarkState();
            bookmarkState.setProvider("com.intellij.ide.bookmark.providers.LineBookmarkProvider");
            bookmarkState.setDescription(element.name());
            bookmarkState.setType(type);
            bookmarkState.getAttributes().put("url", fileDescriptorToUrl(element.fileDescriptor()));
            if (element.linenumber() != -1) {
                bookmarkState.getAttributes().put("line", String.valueOf(element.linenumber()));
            }
            Bookmark bookmark = bookmarksManager.createBookmark(bookmarkState);
            if (bookmark != null) {
                boolean success = group.add(bookmark, type, element.name());
                if (!success) Notify.error("Bookmark add fail, description:[" + element + "]");
            }
        }
    }

    private static String fileDescriptorToUrl(String fileDescriptor) {
        // in jar file?
        if (fileDescriptor.contains(".jar!")) {
            return "jar://" + fileDescriptor;
        }
        return "file://" + fileDescriptor;
    }

    private static String urlToFileDescriptor(String url) {
        File file = new File(url);
        if (file.exists()) {
            return file.getPath();
        }
        if (url.startsWith("jar://")) {
            return url.substring(6);
        }
        if (url.startsWith("file://")) {
            return url.substring(7);
        }
        return url;
    }
}

File: src/main/java/io/github/hhy/bookmark/manager/HighVersionBookManager.kt
package io.github.hhy.bookmark.manager

import com.intellij.ide.bookmark.*
import com.intellij.openapi.project.Project
import io.github.hhy.bookmark.element.BookmarkElement
import io.github.hhy.bookmark.element.Element
import io.github.hhy.bookmark.element.GroupElement
import io.github.hhy.bookmark.util.FDUtil
import java.io.File

class HighVersionBookManager : MyBookmarkManager {

    override fun getAllBookmarks(project: Project): Map<String, Map<String, BookmarkElement>> {
        val bookmarkManager = project.getService(BookmarksManager::class.java)
        if (bookmarkManager is BookmarksManagerImpl) {
            return bookmarkManager.state.groups.associate { group ->
                group.name to group.bookmarks.associate {
                    val attr = it.attributes
                    val ele = Element.withBookmark(
                        attr["url"]!!.let(HighVersionBookManager::urlToFileDescriptor),
                        attr["line"]?.let(Integer::parseInt) ?: -1,
                        it.description ?: "",
                        group.name,
                        it.type.toString()
                    )
                    ele.key() to ele
                }
            }
        }
        return emptyMap()
    }

    override fun addGroup(project: Project, ele: GroupElement): BookmarkGroup {
        val bookmarksManager = project.getService(BookmarksManager::class.java)
        return bookmarksManager.getGroup(ele.name) ?: return bookmarksManager.addGroup(ele.name, false)!!
    }

    override fun addBookmark(project: Project, ele: BookmarkElement): Bookmark {
        val bookmarksManager = project.getService(BookmarksManager::class.java)
        val group: BookmarkGroup = addGroup(project, Element.withGroup(ele.group))
        val type: BookmarkType = BookmarkType.valueOf(ele.bookmarkType)
        val bookmarkState = BookmarkState().also {
            it.provider = "com.intellij.ide.bookmark.providers.LineBookmarkProvider"
            it.description = ele.name
            it.type = type
            it.attributes["url"] = fileDescriptorToUrl(ele.fileDescriptor)
            if (ele.linenumber != -1) {
                it.attributes["line"] = java.lang.String.valueOf(ele.linenumber)
            }
        }
        val bookmark = bookmarksManager.createBookmark(bookmarkState)
        return bookmark?.also {
            group.add(bookmark, type, ele.name)
        } ?: throw RuntimeException()
    }

    override fun removeInvalid(project: Project): List<Bookmark> {
        val invalid: MutableList<Bookmark> = ArrayList()
        val bookmarksManager = project.getService(BookmarksManager::class.java)
        for (bookmark in bookmarksManager.bookmarks) {
            if (bookmark::class.java.name.contains("InvalidBookmark")) {
                bookmarksManager.remove(bookmark)
                invalid += bookmark
            }
        }
        return invalid
    }

    companion object {
        @JvmStatic
        fun urlToFileDescriptor(url: String): String {
            val file = File(url)
            if (file.exists()) {
                return file.path
            }
            var url = if (url.startsWith("jar://")) {
                url.substring(6)
            } else if (url.startsWith("file://")) {
                url.substring(7)
            } else {
                url
            }
            return FDUtil.formatSeparator(url)
        }

        @JvmStatic
        private fun fileDescriptorToUrl(fileDescriptor: String): String {
            // in jar file?
            return if (fileDescriptor.contains(".jar!")) {
                "jar://$fileDescriptor"
            } else "file://$fileDescriptor"
        }
    }
}
